### PR TITLE
rgblight support for ps2avrGB

### PR DESCRIPTION
**Pull request by luizribeiro**
This adds support for RGB underglow on ps2avrGB boards. It includes a few custom functions that can be mapped to control RGB modes, etc. But these aren't used in the default keymap right now.

I wasn't able to reuse the I2C implementation under QMK core (since ps2avrGB uses V-USB and quite different registers for setup), so I ended up just merging my own I2C implementation in. Maybe it's possible to reuse it, but I didn't have the time to spend on this so it feels like a decent first step (it also minimizes the changes to QMK core, which is a good idea IMO).

Because of this, I had to stop including `light_ws2812.c` on ps2avrGB builds. This meant I had to introduce a new build option, called `DISABLE_WS2812`.

### CODE REVIEW

**Comment by TerryMathews** [comment]
So... this is an interesting problem, and one that might make us need to rethink how RGB LED support is implemented in QMK. I'll paraphrase for anyone else who reads this comment:

To date, every RGB setup added to QMK has operated under one assumption: that the WS2812B LEDs would be directly driven by the MCU. **All of the backing code assumes that.** The BootMapper-based boards and the Jigon-based boards (which have working QMK/TMK ports) however use an ATtiny84 to drive the LEDs. It is communicated with by I2C (on both boards) and works in largely the same fashion although Jigon uses slightly different commands than BootMapper to control the ATtiny84. They're both fairly well documented.

Now that someone has the BootMapper RGB LEDs working in QMK, it might be a good time to decouple the RGB implementation code from the communication code.

**Comment by luizribeiro** [comment]
I actually think that the code for driving the WS2812B with the MCU is fairly isolated already. The only function that relies on it is `rgblight_set`, which is already declared with `__weak__`. So all I had to do was to add a flag to make sure that the code for driving the WS2812B was left out of the build, since the `rgblight_set` function for this keyboard doesn't rely on it.

What exactly are you proposing here?

**Comment by TerryMathews** [comment]
Well, specifically I was thinking about the fact that all the code to date in QMK assumes that RGBLIGHT_ENABLE controls all aspects of RGB lighting. For example, I know of at least on keymap that has two different FN layers depending on whether or not RGBLIGHT is enabled or not.

I think we should leave RGBLIGHT to be the master control for all RGB lighting support and use #defines to switch from ws2812 to attiny84. Which would mean implementing proper ifdefs in the relevant support files.

**Comment by luizribeiro** [comment]
So how about we add a `RGBLIGHT_CUSTOM_DRIVER` flag? This would stop including ws2812-related files on build-time (just like this) and strip the `rgblight_set` implementation out of QMK core. This way keyboards that use this new flag would have to implement `rgblight_set` themselves.

This is very similar to this, but it would turn into a build error if you use the `RGBLIGHT_CUSTOM_DRIVER` flag and don't implement your own `rgblight_set` function.

What do you think?

**Comment by TerryMathews** [comment]
Seems consistent with how we handle custom matrix.c files today. I like it.

**Comment by jackhumbert** [comment]
I'm a fan :)

**Comment by luizribeiro** [comment]
Done! I had to move some of the types under `quantum/light_ws2812.h` into a new file, `quantum/rgblight_types.h` so that we could reuse them even when we're using a custom driver. Happy to consider other options there you guys may have in mind.

File: build_keyboard.mk [hunk 1]
```diff
@@ -198,10 +198,14 @@ endif
...
     OPT_DEFS += -DRGBLIGHT_ENABLE
-    SRC += $(QUANTUM_DIR)/light_ws2812.c
     SRC += $(QUANTUM_DIR)/rgblight.c
     CIE1931_CURVE = yes
     LED_BREATHING_TABLE = yes
+ifeq ($(strip $(RGBLIGHT_CUSTOM_DRIVER)), yes)
+			OPT_DEFS += -DRGBLIGHT_CUSTOM_DRIVER
+else
+			SRC += $(QUANTUM_DIR)/light_ws2812.c
+endif
```
Comment (fredizzimo) [inline build_keyboard.mk]: There's a mix of spaces and tabs here, the rest of the file uses spaces, so could you please chang it to spaces?

**Comment by fredizzimo** [comment]
Is there any reason why you don't use the standard RGB light keys?
```
RGB_TOG,
RGB_MOD,
RGB_HUI,
RGB_HUD,
RGB_SAI,
RGB_SAD,
RGB_VAI,
RGB_VAD
```

It would be nice if we handle the RGB light in the same way everywhere.

I think the I2C stuff should be moved to some shared QMK code, because right now we have the same code duplicated in several places. But that can be done later in another pull request, possibly by someone else.

**Comment by luizribeiro** [comment]
Ha, I had no idea those macro existed. Let me change this, thanks for the comments!

**Comment by Sebb767** [comment]
I'd like to get this working in order to get all my keyboards on qmk.

What's the status of this issue? Is something missing and can I help in some way?

**Comment by luizribeiro** [comment]
I've been using this on my B.mini X2 for the past few months.

I believe we just need to get another round of review on the changes and
fix any merge conflicts.

If folks have any comments on the implementation details I can probably
work on them and fix the merge conflicts this weekend.

File: keyboards/ps2avrGB/matrix.c [hunk 2]
```diff
@@ -93,6 +93,8 @@ uint8_t matrix_scan(void) {
         }
     }
 
+    matrix_scan_user();
```
Comment (jackhumbert) [inline keyboards/ps2avrGB/matrix.c]: This should actually be calling `matrix_scan_quantum()` - did you see this anywhere else? There might be some bad examples of this :)

**Comment by jackhumbert** [comment]
I think this is looking good! My only feedback would be to try to use lufa's built-in I2C/TWI rather than having the i2c driver in the keyboard directory. The include for this is:

    #include "LUFA/Drivers/Peripheral/TWI.h"

This is something we'll be trying to convert everything (lets_split, etc) to down the road!

**Comment by luizribeiro** [comment]
> My only feedback would be to try to use lufa's built-in I2C/TWI rather than having the i2c driver in the keyboard directory.

Is that compatible with V-USB? I think we may not be including anything from LUFA when building V-USB based keyboards (since those microcontrollers aren't even [supported by LUFA](http://www.fourwalledcubicle.com/files/LUFA/Doc/120219/html/_page__a_v_r8_support.htmlhttp://www.fourwalledcubicle.com/files/LUFA/Doc/120219/html/_page__a_v_r8_support.html)).

**Comment by jackhumbert** [comment]
Ah, that's a good point - I'm not sure :) no worries then - we'll (try to) convert it when we do everything else I suppose!

**Comment by jackhumbert** [comment]
Merged in via [this commit](https://github.com/qmk/qmk_firmware/commit/5ad5c230d6cf9f90a8337ec54dee8b14e332d3c6)! Had messy conflicts so it had to be done manually :/

**Pull request merged**